torch/csrc/autograd/saved_variable.cpp

```diff
@@ -144,7 +144,16 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
                 : grad_fn_;
 
   if (!is_leaf_ && !grad_fn) {
-    TORCH_INTERNAL_ASSERT(saved_for, "No grad_fn for non-leaf saved tensor");
+    // This issue was introduced when we added logic to save the original
+    // because now we rely on data_.grad_fn(), but can be unreliable if the
+    // autograd_meta of that saved tensor is cleared with an in-place detach.
+    // As a simple fix, we choose to disallow that behavior here even though
+    // it makes behavior inconsistent depending on whether you are saving
+    // input or output.
+    TORCH_CHECK(
+        saved_for,
+        "Trying to use a saved tensor that has been detached in-place, i.e. with .detach_()."
```

Comment:
btw is inplace detach() bumping the version counter? If so, then the vc check below would also fail. Not sure which of the two error is best...
I would vote for to keep the more precise one here about inplace detach but I'm open.

Reply:
From looking at its kernel in VariableTypeManual, it doesn't redispatch to ADInplaceOrView, so we should be good.

```
>>> import torch
>>> a = torch.tensor(1., requires_grad=True)
>>> a._version
0
>>> a.detach_()
tensor(1.)
>>> a._version
0
>>> a.add_(1)
tensor(2.)
>>> a._version
1
>>>
```